_posts/2020-06-14-dynamic-partition-pruning.md

```diff
@@ -0,0 +1,159 @@
+optimization rules to reorder joins and choose join distribution type have been run. This ensures that dynamic filters do not
+interfere in cost estimations and cause a sub-optimal plan to be picked by the CBO.
+The `PredicatePushDown` rule can end up pushing dynamic filters to unsupported places in the plan via inferencing. 
+This was solved by adding the 
+[`RemoveUnsupportedDynamicFilters`](https://github.com/prestosql/presto/blob/master/presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/RemoveUnsupportedDynamicFilters.java)
+optimizer rule which is responsible for ensuring that
+* Dynamic filters are present only right above a `TableScan` node and only if the subtree is on the probe side of some downstream `JoinNode`
+* Dynamic filters are removed from `JoinNode` if there is no consumer for it on its probe side subtree.
+
+We also run [`DynamicFiltersChecker`](https://github.com/prestosql/presto/blob/master/presto-main/src/main/java/io/prestosql/sql/planner/sanity/DynamicFiltersChecker.java)
+at the end of the planning phase to verify that dynamic filters from a `JoinNode` are consumed by a `FilterNode` above a `TableScan` node
+on it's probe-side subtree.
+ 
+We reuse the existing [`DynamicFilterSourceOperator`](https://github.com/prestosql/presto/blob/master/presto-main/src/main/java/io/prestosql/operator/DynamicFilterSourceOperator.java)
+in `LocalExecutionPlanner` to collect build-side values from each inner join on each worker node. In addition to passing the collected `TupleDomain`
+to [`LocalDynamicFiltersCollector`](https://github.com/prestosql/presto/blob/master/presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFiltersCollector.java) 
+within the same worker node for use in broadcast join probe-side scans, we also pass them to `TaskContext` to populate task 
+status updates for the coordinator. 
+
+`ContinuousTaskStatusFetcher` on the coordinator node pulls task status updates from all worker nodes every 
+`task.status-refresh-max-wait` seconds (default is 1 second). `DynamicFilterService` on the coordinator regularly polls 
+for dynamic filters from task status updates through `SqlQueryExecution` and provides an interface to supply dynamic 
+filters when they are ready. The `ConnectorSplitManager#getSplits` API has been updated to optionally utilize dynamic 
+filters supplied by the `DynamicFilterService`.
+
+In the Hive connector, `BackgroundHiveSplitLoader` can apply dynamic filtering by either completely skipping the listing
+of files within a partition, or by avoiding the creation of splits within a loaded partition if the dynamic filters 
+become available in `InternalHiveSplitFactory#createInternalHiveSplit` due to lazy enumeration of splits.
+
+# Benchmarks
+
+We ran TPC-DS queries on 5 worker nodes cluster of r4.8xlarge machines using TPC-DS scale 1000 data stored in ORC format.
```

Comment:
`using TPC-DS scale 1000 data stored in ORC`

Scales are different for queries to make sure query runs reasonable time (we need to adjust these since some queries are too fast now).

Maybe something like:
```
We ran TPC-DS queries on 5 worker nodes clusters of r4.8xlarge machines using TPC-DS
(scale depends on query to ensure reasonable query runtime) with data stored in ORC format.
```
or just skip scale part

Reply:
Removed mention of scale